kf_lib_data_ingest/etl/ingest_pipeline.py

```diff
@@ -151,20 +154,89 @@ def run(self):
         exception handler so that exceptions are logged.
         """
         self.logger.info('BEGIN data ingestion.')
+        self.stage_outputs = {}
+        self.stage_discovery_sources = {}
+        self.stage_discovery_links = {}
+
         # Top level exception handler
         # Catch exception, log it to file and console, and exit
         try:
             # Iterate over stages and execute them
             output = None
-            for s in self._iterate_stages():
+            for stage in self._iterate_stages():
                 if not output:  # First stage gets no input
-                    output, checks_passed, accounting_data = s.run()
+                    output, concept_discovery_dict = stage.run()
                 else:
-                    output, checks_passed, accounting_data = s.run(output)
+                    output, concept_discovery_dict = stage.run(output)
+
+                # Collapse stage subtypes to their bases for storage
+                # (e.g. GuidedTransformStage -> TransformStage)
+                stage_type = type(stage)
+                while stage_type.__base__ != IngestStage:
+                    stage_type = stage_type.__base__
+
+                write_json(
+                    concept_discovery_dict,
+                    os.path.join(
+                        self.ingest_output_dir,
+                        stage_type.__name__ + '_concept_discovery.json'
+                    )
+                )
+
+                self.stage_outputs[stage_type] = output
+                self.validate_stage_counts(
+                    stage_type, stage.logger, concept_discovery_dict
+                )
         except Exception as e:
             self.logger.exception(str(e))
             self.logger.info('Exiting.')
             exit(1)
 
         # Log the end of the run
         self.logger.info('END data ingestion')
+
+    def validate_stage_counts(
+        self, stage_type, logger, concept_discovery_dict
+    ):
+        """
+        Do some standard stage discovery tests
+        """
+
+        logger.info("Begin Basic Analysis")
+
+        passed_all = True
```

Comment:
This is a little hard to read. Something like this is more readable to me

```suggestion
        # Discovery data is missing
        if not concept_discovery_dict:
            logger.info('Discovery Data Not Found ❌')
            return False

        if not concept_discovery_dict.get('sources'):
            logger.info('Discovery Data Sources Not Found ❌')
            return False

        # Do stage counts validation
        passed_all = True

        passed, message = stage_analyses.check_counts(
            concept_discovery_dict.get('sources'),
            self.data_ingest_config.expected_counts
        )
        passed_all = passed_all and passed
        logger.info(message)

        # Compare stage counts to make sure we didn't lose concepts/concept attributes
        self.stage_discovery_sources[stage_type] = (
            concept_discovery_dict['sources']
        )
        if stage_type == TransformStage:
            if ExtractStage in self.stage_discovery_sources:
                passed, message = stage_analyses.compare_counts(
                    concept_discovery_dict.get('sources'),
                    self.stage_discovery_sources[ExtractStage]
                )
                passed_all = passed_all and passed
                logger.info(message)
            else:
                passed_all = False
                logger.info(
                    'No Extract Discovery Data Sources To Compare ❌'
                )

        logger.info("End Basic Analysis")
        return passed_all
```

Reply:
√